--- third_party/googletest/googletest/include/gtest/internal/gtest-port.h	2023-01-17 16:10:56.252360383 +0800
+++ /home/suze/tp/code/ohos_devtools_trace_resolver/third_party/googletest/googletest/include/gtest/internal/gtest-port.h	2023-01-06 17:58:25.834759489 +0800
@@ -276,7 +276,6 @@
 # include <AvailabilityMacros.h>
 # include <TargetConditionals.h>
 #endif
-
 #include <iostream>  // NOLINT
 #include <locale>
 #include <memory>
@@ -2287,7 +2286,11 @@ using Any = ::absl::any;
 // Otherwise for C++17 and higher use std::any for UniversalPrinter<>
 // specializations.
 #define GTEST_INTERNAL_HAS_ANY 1
-#include <any>
+#undef private
+#define private private
+#include <any>  // NOLINT
+#undef private
+#define private public
 namespace testing {
 namespace internal {
 using Any = ::std::any;
