--- third_party/googletest/googletest/include/gtest/internal/gtest-internal.h	2023-01-17 16:10:56.252360383 +0800
+++ /home/suze/tp/code/ohos_devtools_trace_resolver/third_party/googletest/googletest/include/gtest/internal/gtest-internal.h	2023-01-06 17:58:25.830759482 +0800
@@ -54,7 +54,11 @@
 #include <float.h>
 #include <string.h>
 #include <cstdint>
+#undef private
+#define private private
 #include <iomanip>
+#undef private
+#define private public
 #include <limits>
 #include <map>
 #include <set>
