--- third_party/googletest/googletest/include/gtest/gtest-message.h	2023-01-17 16:10:56.252360383 +0800
+++ /home/suze/tp/code/ohos_devtools_trace_resolver/third_party/googletest/googletest/include/gtest/gtest-message.h	2023-01-06 17:58:25.830759482 +0800
@@ -49,8 +49,11 @@
 
 #include <limits>
 #include <memory>
+#undef private
+#define private private
 #include <sstream>
-
+#undef private
+#define private public
 #include "gtest/internal/gtest-port.h"
 
 GTEST_DISABLE_MSC_WARNINGS_PUSH_(4251 \
